src/theory/arith/arith_rewriter.cpp

```diff
@@ -48,6 +48,70 @@ namespace cvc5::internal {
 namespace theory {
 namespace arith {
 
+
+/**
+ * Flatten a node into a vector of its (direct or indirect) children, collecting
+ * how many times each child occurs in the sum.
+ * A sequence of kinds is given that indicate which kinds to traverse over.
+ * This method is similar to expr::algorithm::flatten but does not use a tree
+ * traversal. Instead it merges subterms, based on counting the number of
+ * occurrences, as a Rational.
+ * @param t The node to be flattened
+ * @param children The resulting list of children
+ * @param kinds A sequence of kinds to consider for flattening
+ */
+template <typename... Kinds>
+bool flattenAndCollectSum(TNode t, std::vector<std::pair<TNode, Rational>>& children, Kinds... kinds)
+{
+  if (!expr::algorithm::canFlatten(t, kinds...))
+  {
+    return false;
+  }
+  std::map<TNode, Rational> countMap;
+  countMap[t] = Rational(1);
+  std::map<TNode, Rational>::iterator it;
+  Kind tk = t.getKind();
+  while (!countMap.empty())
```

Comment:
I don't think that this will entirely avoid tree traversals. There is no guarantee that a node is not visited more than once since there is no global cache.

For example,  for `(+ (+ t 1) (+ (+ t 3) 2))` it will traverse `t` twice. If t is again a sum, you will also traverse its children again.

Reply:
I believe it does, based on the use of an ordered map where nodes that are created later are processed first.  This guarantees that `(+ t 1) and `(+ t 3)` are processed before `t`, and hence collect into a single entry before `t` is processed.

This is a bit subtle. I would be open to suggestions to making this invariant clearer.

Reply:
Ah yes, the ordered processing should ensure that. Maybe add a comment that the ordered map and always processing the last element ensures that it will only process nodes when all of its parents have been processed first.

Reply:
Added a comment